Add Installation and Configuration for Benchmark section (#4124)
opensearch-project/documentation-website: _benchmark/installing-benchmark.md

---
layout: default
title: Installing OpenSearch Benchmark
nav_order: 5
has_children: false
---

# Installing OpenSearch Benchmark

You can install OpenSearch Benchmark directly on a host running Linux or macOS or you can run OpenSearch Benchmark in a Docker container on any compatible host. This page describes high-level considerations for your OpenSearch Benchmark host, as well as instructions for installing OpenSearch Benchmark.

> ## Docker limitations

Reviewer: Do we want to lead with this as a heading? Can we make it a note instead?

Reviewer: Or maybe add it lower down where you are discussing the Docker install?

Author: I'll lower it down.

Some OpenSearch Benchmark functionality is unavailable when you run OpenSearch Benchmark in a Docker container. Specifically, the following restrictions apply:
- You cannot distribute load worker coordinator hosts.

> - You can only use the `benchmark-only` pipeline.

Reviewer 2: It might be nice to elaborate on what the two of these mean. New-comers might not entirely understand this unless they play around with OSB. We're also planning to rename `load-worker-coordinator-hosts` in the future as it's a mouthful and doesn't offer much insight into what it is. Perhaps something like:
```
- OSB cannot distribute load from multiple hosts (i.e. cannot use load worker coordinator hosts)
- OSB cannot provision OpenSearch nodes and can only run tests on pre-existing clusters (i.e. can only invoke OSB commands using `benchmark-only` pipeline)
```
There's probably a better way to say it but just an idea!

## Choosing appropriate hardware

> OpenSearch Benchmark can be used to provision OpenSearch nodes for testing. If you intend to use OpenSearch Benchmark to provision nodes in your environment, then you will need to install OpenSearch Benchmark directly on each host in the cluster. Additionally, you must configure each host in the cluster for OpenSearch. See [Installing OpenSearch]({{site.url}}{{site.baseurl}}/install-and-configure/install-opensearch/index/) for guidance about important host settings you should review.

Reviewer: ```suggestion
OpenSearch Benchmark can be used to provision OpenSearch nodes for testing. If you intend to use OpenSearch Benchmark to provision nodes in your environment, then install OpenSearch Benchmark directly on each host in the cluster. Additionally, you must configure each host in the cluster for OpenSearch. See [Installing OpenSearch]({{site.url}}{{site.baseurl}}/install-and-configure/install-opensearch/index/) for guidance on important host settings you should review.
```

Remember that OpenSearch Benchmark cannot be used to provision OpenSearch nodes when you run OpenSearch Benchmark in a Docker container. If you want to use OpenSearch Benchmark to provision nodes, or if you want to distribute the benchmark workload with the OpenSearch Benchmark daemon, then you must install OpenSearch Benchmark directly on each host using Python and pip.
{: .important}

> When you select a host, you should also think about which workloads you want to run. To see a list of default benchmark workload, visit the [opensearch-benchmark-workloads](https://github.com/opensearch-project/opensearch-benchmark-workloads) repository on GitHub. As a general rule, make sure that the OpenSearch Benchmark host has enough free storage space to store the compressed data and the fully decompressed data corpus once OpenSearch Benchmark is installed.

Reviewer 2: Nit: `To see a list of default benchmark workload...` --> `To see a list of default benchmark workloads...`

Use [pyenv](https://github.com/pyenv/pyenv) to manage multiple versions of Python on your host. This is especially useful if your "system" version of Python is older than version 3.8.
{: .tip}

- Check that Python 3.8 or newer is installed:

  ```bash
  python3 --version
  ```
  {% include copy.html %}

- Check that `pip` is installed and functions:

  ```bash
  pip --version
  ```
  {% include copy.html %}

- _Optional_ Check that your installed version of `git` is **Git 1.9 or newer**. `git` is not required for OpenSearch Benchmark installation, but it is required to fetch benchmark workload resources from a repository when you want to perform tests. See the official Git [Documentation](https://git-scm.com/doc) for help installing Git. 

  ```bash
  git --version
  ```
  {% include copy.html %}

### Completing the installation

After the prerequisites are installed, you can install OpenSearch Benchmark using the following command:

```bash
pip install opensearch-benchmark
```
{% include copy.html %}

After the installation completes you can use the following command to display help information:

```bash
opensearch-benchmark -h
```
{% include copy.html %}

> OpenSearch Benchmark will exit with an error if no arguments are supplied. Use the `-h` option to print the help text in the command line if you want to test the installation.

Reviewer 2: As of version 0.5.0, OpenSearch-Benchmark will not exit with an error if no arguments are supplied. It will provide the help text instead. See example here:
```
hoangia@3c22fbd0d988 opensearch-benchmark % opensearch-benchmark --version
opensearch-benchmark 0.5.0 (git revision: b98cebd)
hoangia@3c22fbd0d988 opensearch-benchmark % opensearch-benchmark
usage: opensearch-benchmark [-h] [--version]

   ____                  _____                      __       ____                  __                         __
  / __ \____  ___  ____ / ___/___  ____ ___________/ /_     / __ )___  ____  _____/ /_  ____ ___  ____ ______/ /__
 / / / / __ \/ _ \/ __ \\__ \/ _ \/ __ `/ ___/ ___/ __ \   / __  / _ \/ __ \/ ___/ __ \/ __ `__ \/ __ `/ ___/ //_/
/ /_/ / /_/ /  __/ / / /__/ /  __/ /_/ / /  / /__/ / / /  / /_/ /  __/ / / / /__/ / / / / / / / / /_/ / /  / ,<
\____/ .___/\___/_/ /_/____/\___/\__,_/_/   \___/_/ /_/  /_____/\___/_/ /_/\___/_/ /_/_/ /_/ /_/\__,_/_/  /_/|_|
    /_/

 A benchmarking tool for OpenSearch

optional arguments:
  -h, --help  show this help message and exit
  --version   show program's version number and exit

Find out more about Benchmark at https://opensearch.org/docs
```

> OpenSearch Benchmark will exit with an error if no arguments are supplied. Use the `-h` option to print the help text in the command line if you want to test the installation.

Reviewer: Not sure what you mean here. And does it belong above the command?

Author: Removing this line per Ian's comment.

{: .warning}

Now that OpenSearch Benchmark is installed on your host you can learn about [Configuring OpenSearch Benchmark]({{site.url}}{{site.baseurl}}/benchmark/configuring-benchmark/).

## Installing with Docker

You can find the official Docker images for OpenSearch Benchmark on [Docker Hub](https://hub.docker.com/r/opensearchproject/opensearch-benchmark) or on the [Amazon ECR Public Gallery](https://gallery.ecr.aws/opensearchproject/opensearch-benchmark).

To pull the image from Docker Hub:

```bash
docker pull opensearchproject/opensearch-benchmark:latest
```
{% include copy.html %}

To pull the image from Amazon ECR:

```bash
docker pull public.ecr.aws/opensearchproject/opensearch-benchmark:latest
```
{% include copy.html %}

### Running Benchmark with Docker

> To run OpenSearch Benchmark, you use `docker run` to launch a container. OpenSearch Benchmark subcommands are passed as arguments when you start the container. OpenSearch Benchmark then processes the command, then stops the container after the requested operation completes.

Reviewer: ```suggestion
To run OpenSearch Benchmark, use `docker run` to launch a container. OpenSearch Benchmark sub-commands are passed as arguments when you start the container. OpenSearch Benchmark then processes the command, then stops the container after the requested operation completes.
```

> For example, the following command will print the help text for OpenSearch Benchmark to the command line, and then stop the container:

Reviewer: ```suggestion
For example, the following command prints the help text for OpenSearch Benchmark to the command line, and then stops the container:
```

```bash
docker run opensearchproject/opensearch-benchmark -h
```
{% include copy.html %}

> OpenSearch Benchmark will exit with an error if no arguments are supplied. Use the `-h` option to print the help text in the command line if you want to test the installation.

Reviewer: ```suggestion
OpenSearch Benchmark exits with an error if no arguments are supplied. Use the `-h` option to print the help text in the command line if you want to test the installation.
```

Reviewer: Might be clearer to say, "If you want to test the installation, use the 'h' option to display the help text in the command line."

{: .warning}

### Establishing volume persistence in a Docker container

To make sure your benchmark data and logs persist after your Docker container stops, specify a Docker volume to mount to the image when you work with OpenSearch Benchmark.

Use the `-v` option and specify a local directory to mount, and a directory in the container where the volume is attached.

The following example command creates a volume in a user's home directory, mounts the volume to the OpenSearch Benchmark container at `/opensearch-benchmark/.benchmark`, and then executes a test benchmark using the geonames workload. Some client options are also specified:

```bash

> run -v $HOME/benchmarks:/opensearch-benchmark/.benchmark opensearchproject/opensearch-benchmark execute_test --target-hosts https://198.51.100.25:9200 --pipeline benchmark-only --workload geonames --client-options basic_auth_user:admin,basic_auth_password:admin,verify_certs:false --test-mode

Reviewer 2: `/opensearch-benchmark/.benchmark` --> `.benchmark`. See reasoning below

```
{% include copy.html %}

> See [Configuring OpenSearch Benchmark]({{site.url}}{{site.baseurl}}/benchmark/configuring-benchmark/) to learn more about the files and subdirectories located in `/opensearch-benchmark/.benchmark`.

Reviewer 2: `/opensearch-benchmark/.benchmark` can be switched to `~/.benchmark` as we have updated the Docker image to use PyPi and removed the need for `/opensearch-benchmark` venv.

---
layout: default
title: Installing OpenSearch Benchmark
nav_order: 5
has_children: false
---

# Installing OpenSearch Benchmark

> You can install OpenSearch Benchmark directly on a host running Linux or macOS or you can run OpenSearch Benchmark in a Docker container on any compatible host. This page describes high-level considerations for your OpenSearch Benchmark host, as well as instructions for installing OpenSearch Benchmark.

Reviewer 3: ```suggestion
You can install OpenSearch Benchmark directly on a host running Linux or macOS, or you can run OpenSearch Benchmark in a Docker container on any compatible host. This page provides general considerations for your OpenSearch Benchmark host as well as instructions for installing OpenSearch Benchmark.
```

## Choosing appropriate hardware

> OpenSearch Benchmark can be used to provision OpenSearch nodes for testing. If you intend to use OpenSearch Benchmark to provision nodes in your environment, then install OpenSearch Benchmark directly on each host in the cluster. Additionally, you must configure each host in the cluster for OpenSearch. See [Installing OpenSearch]({{site.url}}{{site.baseurl}}/install-and-configure/install-opensearch/index/) for guidance on important host settings you should review.

Reviewer 3: The third sentence is slightly vague. Do we mean something like "in the cluster to work with OpenSearch"?

> OpenSearch Benchmark can be used to provision OpenSearch nodes for testing. If you intend to use OpenSearch Benchmark to provision nodes in your environment, then install OpenSearch Benchmark directly on each host in the cluster. Additionally, you must configure each host in the cluster for OpenSearch. See [Installing OpenSearch]({{site.url}}{{site.baseurl}}/install-and-configure/install-opensearch/index/) for guidance on important host settings you should review.

Reviewer 3: ```suggestion
OpenSearch Benchmark can be used to provision OpenSearch nodes for testing. If you intend to use OpenSearch Benchmark to provision nodes in your environment, then install OpenSearch Benchmark directly on each host in the cluster. Additionally, you must configure each host in the cluster for OpenSearch. See [Installing OpenSearch]({{site.url}}{{site.baseurl}}/install-and-configure/install-opensearch/index/) for guidance on important host settings.
```

Remember that OpenSearch Benchmark cannot be used to provision OpenSearch nodes when you run OpenSearch Benchmark in a Docker container. If you want to use OpenSearch Benchmark to provision nodes, or if you want to distribute the benchmark workload with the OpenSearch Benchmark daemon, then you must install OpenSearch Benchmark directly on each host using Python and pip.
{: .important}

When you select a host, you should also think about which workloads you want to run. To see a list of default benchmark workloads, visit the [opensearch-benchmark-workloads](https://github.com/opensearch-project/opensearch-benchmark-workloads) repository on GitHub. As a general rule, make sure that the OpenSearch Benchmark host has enough free storage space to store the compressed data and the fully decompressed data corpus once OpenSearch Benchmark is installed.

> If you want to benchmark with a default workload, then use the following table to determine the approximate minimum amount of required free space needed by adding the compressed size with the uncompressed size:

Reviewer 3: "amount of free space required by adding the compressed size to the uncompressed size"? (required and needed may be redundant)

> If you want to benchmark with a default workload, then use the following table to determine the approximate minimum amount of required free space needed by adding the compressed size with the uncompressed size:

Reviewer 3: ```suggestion
If you want to benchmark with a default workload, then use the following table to determine the approximate minimum amount of required free space needed by adding the compressed size with the uncompressed size.
```

> | Workload Name | Document Count | Compressed Size | Uncompressed Size |

Reviewer 3: ```suggestion
| Workload name | Document count | Compressed size | Uncompressed size |
```

| :----: | :----: | :----: | :----: |
| eventdata | 20,000,000 | 756.0 MB | 15.3 GB |
| geonames | 11,396,503 | 252.9 MB | 3.3 GB |
| geopoint | 60,844,404 | 482.1 MB | 2.3 GB |
| geopointshape | 60,844,404 | 470.8 MB | 2.6 GB |
| geoshape | 60,523,283 | 13.4 GB | 45.4 GB |
| http_logs | 247,249,096 | 1.2 GB | 31.1 GB |
| nested | 11,203,029 | 663.3 MB | 3.4 GB |
| noaa | 33,659,481 | 949.4 MB | 9.0 GB |
| nyc_taxis | 165,346,692 | 4.5 GB | 74.3 GB |
| percolator | 2,000,000 | 121.1 kB | 104.9 MB |
| pmc | 574,199 | 5.5 GB | 21.7 GB |
| so | 36,062,278 | 8.9 GB | 33.1 GB |

> Your OpenSearch Benchmark host should use solid-state drives (SSDs) for storage because they perform significantly faster at read and write operations than traditional spinning-disk hard drives. Spinning-disk hard drives can introduce performance bottlenecks, which can make benchmark results unreliable and inconsistent.

Reviewer 3: ```suggestion
Your OpenSearch Benchmark host should use solid-state drives (SSDs) for storage because they perform read and write operations significantly faster than traditional spinning-disk hard drives. Spinning-disk hard drives can introduce performance bottlenecks, which can make benchmark results unreliable and inconsistent.
```

{: .tip}

## Installing on Linux and macOS

> If you want to run OpenSearch Benchmark in a Docker container, see [Installing with Docker](#installing-with-docker). The OpenSearch Benchmark Docker image includes all of the required software, so there are no additional steps for you to take.

Reviewer 3: ```suggestion
If you want to run OpenSearch Benchmark in a Docker container, see [Installing with Docker](#installing-with-docker). The OpenSearch Benchmark Docker image includes all of the required software, so there are no additional steps required.
```

{: .important}

> To install OpenSearch Benchmark directly on a UNIX host such as Linux or macOS, make sure you have **Python 3.8 or newer** installed. 

Reviewer 3: ```suggestion
To install OpenSearch Benchmark directly on a UNIX host, such as Linux or macOS, make sure you have **Python 3.8 or later** installed. 
```

If you need help installing Python, refer to the official [Python Setup and Usage](https://docs.python.org/3/using/index.html) documentation. 

### Checking software dependencies

> Before you being installing OpenSearch Benchmark, check the following software dependencies.

Reviewer 3: ```suggestion
Before you begin installing OpenSearch Benchmark, check the following software dependencies.
```

> Use [pyenv](https://github.com/pyenv/pyenv) to manage multiple versions of Python on your host. This is especially useful if your "system" version of Python is older than version 3.8.

Reviewer 3: ```suggestion
Use [pyenv](https://github.com/pyenv/pyenv) to manage multiple versions of Python on your host. This is especially useful if your "system" version of Python is earlier than version 3.8.
```

{: .tip}

> - Check that Python 3.8 or newer is installed:

Reviewer 3: ```suggestion
- Check that Python 3.8 or later is installed:
```

  ```bash
  python3 --version
  ```
  {% include copy.html %}

> - Check that `pip` is installed and functions:

Reviewer 3: ```suggestion
- Check that `pip` is installed and functional:
```

  ```bash
  pip --version
  ```
  {% include copy.html %}

- _Optional_ Check that your installed version of `git` is **Git 1.9 or newer**. `git` is not required for OpenSearch Benchmark installation, but it is required to fetch benchmark workload resources from a repository when you want to perform tests. See the official Git [Documentation](https://git-scm.com/doc) for help installing Git. 

>   ```bash

Reviewer 3: Reference the example in the text that precedes it.

  git --version
  ```
  {% include copy.html %}

### Completing the installation

> After the prerequisites are installed, you can install OpenSearch Benchmark using the following command:

Reviewer 3: "After the required software is installed"?

```bash
pip install opensearch-benchmark
```
{% include copy.html %}

> After the installation completes you can use the following command to display help information:

Reviewer 3: ```suggestion
After the installation completes, you can use the following command to display help information:
```

```bash
opensearch-benchmark -h
```
{% include copy.html %}

> Now that OpenSearch Benchmark is installed on your host you can learn about [Configuring OpenSearch Benchmark]({{site.url}}{{site.baseurl}}/benchmark/configuring-benchmark/).

Reviewer 3: ```suggestion
Now that OpenSearch Benchmark is installed on your host, you can learn about [Configuring OpenSearch Benchmark]({{site.url}}{{site.baseurl}}/benchmark/configuring-benchmark/).
```

## Installing with Docker

You can find the official Docker images for OpenSearch Benchmark on [Docker Hub](https://hub.docker.com/r/opensearchproject/opensearch-benchmark) or on the [Amazon ECR Public Gallery](https://gallery.ecr.aws/opensearchproject/opensearch-benchmark).

### Docker limitations

Some OpenSearch Benchmark functionality is unavailable when you run OpenSearch Benchmark in a Docker container. Specifically, the following restrictions apply:

- OpenSearch Benchmark cannot distribute load from multiple hosts, such as load worker coordinator hosts.
- OpenSearch Benchmark cannot provision OpenSearch nodes and can only run tests on previously existing clusters. You can only invoke OpenSearch Benchmark commands using the `benchmark-only` pipeline.

> ### Pulling the the Docker images

Reviewer 3: ```suggestion
### Pulling the Docker images
```

> To pull the image from Docker Hub:

Reviewer 3: Let's complete the sentence: "To pull the image from Docker Hub, run the following..."

```bash
docker pull opensearchproject/opensearch-benchmark:latest
```
{% include copy.html %}

> To pull the image from Amazon ECR:

Reviewer 3: Same

> To pull the image from Amazon ECR:

Reviewer 3: ```suggestion
To pull the image from Amazon Elastic Container Registry (Amazon ECR):
```

```bash
docker pull public.ecr.aws/opensearchproject/opensearch-benchmark:latest
```
{% include copy.html %}

### Running Benchmark with Docker

> To run OpenSearch Benchmark, use `docker run` to launch a container. OpenSearch Benchmark sub-commands are passed as arguments when you start the container. OpenSearch Benchmark then processes the command, then stops the container after the requested operation completes.

Reviewer 3: ```suggestion
To run OpenSearch Benchmark, use `docker run` to launch a container. OpenSearch Benchmark subcommands are passed as arguments when you start the container. OpenSearch Benchmark then processes the command and stops the container after the requested operation completes.
```

> For example, the following command prints the help text for OpenSearch Benchmark to the command line, and then stops the container:

Reviewer 3: ```suggestion
For example, the following command prints the help text for OpenSearch Benchmark to the command line and then stops the container:
```

```bash
docker run opensearchproject/opensearch-benchmark -h
```
{% include copy.html %}

### Establishing volume persistence in a Docker container

To make sure your benchmark data and logs persist after your Docker container stops, specify a Docker volume to mount to the image when you work with OpenSearch Benchmark.

> Use the `-v` option and specify a local directory to mount, and a directory in the container where the volume is attached.

Reviewer 3: ```suggestion
Use the `-v` option to specify a local directory to mount and a directory in the container where the volume is attached.
```

> Use the `-v` option and specify a local directory to mount, and a directory in the container where the volume is attached.

Reviewer 3: "to which" instead of "where"?

> The following example command creates a volume in a user's home directory, mounts the volume to the OpenSearch Benchmark container at `/opensearch-benchmark/.benchmark`, and then executes a test benchmark using the geonames workload. Some client options are also specified:

Reviewer 3: ```suggestion
The following example command creates a volume in a user's home directory, mounts the volume to the OpenSearch Benchmark container at `/opensearch-benchmark/.benchmark`, and then runs a test benchmark using the geonames workload. Some client options are also specified:
```
